gandalf.js

```diff
@@ -6,39 +6,39 @@
  *
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?